spectrum-network/src/protocol_handler.rs

```diff
@@ -32,77 +50,254 @@ pub enum ProtocolHandlerError {
+                                } else {
+                                    self.behaviour.inject_malformed_mesage(
+                                        peer_id,
+                                        MalformedMessage::VersionMismatch {
+                                            negotiated_ver,
+                                            actual_ver,
+                                        },
+                                    )
+                                }
+                            }
+                            Some(Err(_)) => self
+                                .behaviour
+                                .inject_malformed_mesage(peer_id, MalformedMessage::UnknownFormat),
+                            None => self.behaviour.inject_protocol_requested(peer_id, None),
+                        }
+                    }
+                    ProtocolEvent::RequestedLocal(peer_id) => {
+                        self.behaviour.inject_protocol_requested_locally(peer_id);
+                    }
+                    ProtocolEvent::Enabled {
+                        peer_id,
+                        protocol_ver: negotiated_ver,
+                        sink,
+                        handshake,
+                    } => {
+                        self.peers.insert(peer_id, sink);
+                        match handshake.map(
+                            codec::decode::<
+                                <<TBehaviour as ProtocolBehaviour>::TProto as ProtocolSpec>::THandshake,
+                            >,
+                        ) {
+                            Some(Ok(hs)) => {
+                                let actual_ver = hs.version();
+                                if actual_ver == negotiated_ver {
+                                    self.behaviour.inject_protocol_enabled(peer_id, Some(hs));
+                                } else {
+                                    self.behaviour.inject_malformed_mesage(
+                                        peer_id,
+                                        MalformedMessage::VersionMismatch {
+                                            negotiated_ver,
+                                            actual_ver,
+                                        },
+                                    )
+                                }
+                            }
+                            Some(Err(_)) => self
+                                .behaviour
+                                .inject_malformed_mesage(peer_id, MalformedMessage::UnknownFormat),
+                            None => self.behaviour.inject_protocol_enabled(peer_id, None),
+                        }
+                    }
+                    ProtocolEvent::Disabled(peer_id) => {
+                        self.behaviour.inject_protocol_disabled(peer_id);
+                    }
+                }
+                continue;
+            }
+
+            return Poll::Pending;
```

Comment:
Don't we need to signal the waker in `cx` every time we return `Poll::Pending`?

Reply:
I've looked at how poll is implemented in libp2p components and haven't noticed them using this practice. Can you elaborate why it's needed?

Reply:
It turns out that it's often not needed. If we're implementing a 'leaf' future that doesn't depend on other existing futures, we'd need to call the waker like in https://github.com/libp2p/rust-libp2p/blob/1b793242e6e31c1556469518d2fea4675a5fe8af/muxers/mplex/src/io.rs#L229 for instance.

But that's not really happening with many libp2p examples including ours. FWIW they have a coding guideline on this topic https://github.com/libp2p/rust-libp2p/blob/1b793242e6e31c1556469518d2fea4675a5fe8af/muxers/mplex/src/io.rs#L229